src/misc/host_locale_macos.cpp

```diff
@@ -179,6 +405,198 @@ static std::string get_language_file(std::string& log_info)
+		return {};
+	}
+
+	// We expect the root of the file to be a dictionary
+	if (CFGetTypeID(plist_ref) != CFDictionaryGetTypeID()) {
+		CFRelease(plist_ref);
+		return {};
+	}
+
+	// Find the array containing input sources
+	const auto root_ref = static_cast<CFDictionaryRef>(plist_ref);
+	if (!CFDictionaryContainsKey(root_ref, KeyMain)) {
+		CFRelease(plist_ref);
+		return {};
+	}
+	const auto array_ptr = CFDictionaryGetValue(root_ref, KeyMain);
+	if (!array_ptr || CFGetTypeID(array_ptr) != CFArrayGetTypeID()) {
+		CFRelease(plist_ref);
+		return {};
+	}
+	const auto array_ref = static_cast<CFArrayRef>(array_ptr);
+
+	// Loop over all the input sources
+	const auto sources_count = CFArrayGetCount(array_ref);
+	for (auto index = 0; index < sources_count; ++index) {
+		// NOTE: this code does not recognize logographic (Chinese,
+		// Korean, Japanese, etc.) writing systems nor transliteration
+		// input sources; their configuration is much more complex,
+		// plus we do not support these alphabets anyway.
+
+		// Make sure the array element is a dictionary, skip otherwise
+		const auto entry_ptr = CFArrayGetValueAtIndex(array_ref, index);
+		if (!entry_ptr || CFGetTypeID(entry_ptr) != CFDictionaryGetTypeID()) {
+			continue;
+		}
+		const auto entry_ref = static_cast<CFDictionaryRef>(entry_ptr);
+
+		// Make sure the entry is a keyboard layout, skip otherwise
+		if (!CFDictionaryContainsKey(entry_ref, KeySourceKind)) {
+			continue;
+		}
+		const auto source_kind_ptr = CFDictionaryGetValue(entry_ref, KeySourceKind);
+		if (!source_kind_ptr || CFGetTypeID(source_kind_ptr) != CFStringGetTypeID()) {
+			continue;		
+		}
+		const auto source_kind_ref = static_cast<CFStringRef>(source_kind_ptr);
+		if (to_string(source_kind_ref) != keyboardLayoutSourceKind) {
+			continue;
+		}
+
+		// Retrieve keyboard layout name and identifier
+		if (!CFDictionaryContainsKey(entry_ref, KeyLayoutName) ||
+		    !CFDictionaryContainsKey(entry_ref, KeyLayoutId)) {
+			continue;
+		}
+		const auto layout_name_ptr = CFDictionaryGetValue(entry_ref, KeyLayoutName);
+		const auto layout_id_ptr   = CFDictionaryGetValue(entry_ref, KeyLayoutId);
+		if (!layout_name_ptr || !layout_id_ptr ||
+		    CFGetTypeID(layout_name_ptr) != CFStringGetTypeID() ||
+		    CFGetTypeID(layout_id_ptr) != CFNumberGetTypeID()) {
+			continue;
+		}
+		const auto layout_id_ref  = static_cast<CFNumberRef>(layout_id_ptr);
+		if (CFNumberIsFloatType(layout_id_ref) ||
+		    CFNumberGetByteSize(layout_id_ref) > MaxIntSize) {
+		    	continue;
+		}
+		const auto layout_name_ref = static_cast<CFStringRef>(layout_name_ptr);
+
+		// Convert keyboard layout name and identifier to C++ types
+		const auto layout_name = to_string(layout_name_ref);
+		int64_t layout_id = 0;
+		CFNumberGetValue(layout_id_ref, kCFNumberSInt64Type, &layout_id);
+
+		apple_layouts.emplace_back(layout_id, layout_name);
+	}
+
+	CFRelease(plist_ref);
+	return get_layouts_maybe_codepages(apple_layouts, log_info);
+}
```

Comment:
Apple has created a hideous disaster with their C/C++ Core Foundation APIs. It has to be deliberate given their otherwise minimal/elegant aesthetic; probably to maximise harm to developers using cross-platform C++ and STL, with the escape hatch to their Swift language. :facepalm:

Reply:
It's, IMHO, not that bad - it's probably just meant to be used in some kind of wrapper classes rather than directly.

At least the API is consistent, doesn't force me to use UTF-16, or anything like this.

Reply:
@interloper98 Huh? CoreFoundation is well designed and well documented, and it's not really meant to be used directly. C/C++ programmers using it directly for _entire programs_ kinda deserve it, yeah...

All the "ugliness" is due to the limitations of C.

Even C++ is generally a horrific choice for GUI programming. The sensible choice is exactly what Apple has been proposing for decades: only use C for really low system level interfacing, C++ for the really performance intensive, algorithm heavy stuff, then the rest should be done in a pleasant high level language, preferably dynamic, such as ObjC or Swift.

Game coders have been doing the same things for decades (C++ engine controlled by Lua or C# code), many the Adobe products do the same (e.g., Lightroom UI is in Lua), and even non-sucky FOSS projects such as Blender (rendering backend is C++, UI logic is Python).

Reply:
Yeah, because as i said, it's not meant to be used directly. You don't need "absolute control over memory allocations" for GUI apps. Maybe some low-level embedded UI-lesz software in your car, yes, but not your desktop app. Then again, really mission critical software doesn't normally use dynamic memory management because you cannot reason about memory fragmentation without a GC. Full manual memory management can be _less_ safe than a good GC because if your malloc in your space shuttle control software gives you an error due to memory fragmentation, it's game ovet...

The CF ref counting function are meant to be used by rhe Swift and ObjC stdlibs and SDKs and that's it.

Reply:
I don't think such things would be hard for people capable designing entire OSes and ecosystems. They did not do it becaus first-class support for writing modern macOS apps in C and C++ was never a top priority or even design consideration.

It's not a question of capability; I'm capable of doing many, many things that I choose not to do 😄 Same goes for every individual and company.

CF is a bridging library. You can read about it's interesting histor here, but it was build as a bridge to aid Cocoa development in ObjC.

https://www.reddit.com/r/swift/comments/40eswd/swift_standard_library_versus_core_foundation/

Let's just conclude this by saying that handling dynamic maps in C is a nightmare, no matter what you do 😄 Same for strings and any data structure really that's not a static one-dimensional array of primitive types 😆

Reply:
I don't think Apple has ever done C++ APIs.  If you look at the SDL source code, they write their Mac platform code in Objective-C and I assume they have a Swift API as well.

This just looks like a thing you would use to write bindings for another language.  Basically every language has a C FFI so you always need something like this to bridge languages.  I've seen this kind of thing with C++ libraries as well.  Exposing higher level types to C always looks ugly but it's really the only way to make an API that can be portable to other languages.

Reply:
Nice fine, but I'd just use ObjC like SDL. Introducing 3rd party deps like that is never a good idea if you only need it for a few use cases. I'd rather live with 1000 lines of ugly direct CF code than another layer of abstraction, honestly, and it's questionable how long they're gonna support this, etc.

Reply:
For now I can think of only one other direct use of macOS API - physical CD (including audio tracks) support. I intend to investigate this, I have already found some open source code using IOKit for CD access.

But I don’t intend to use/create any generic bindings. Maybe the `CFString` to `std::string` conversion routine will go to `cross.h`, but that’s really it.